File: Frontend/smartdss-frontend/src/services/financeService.ts
import api from './api';
import type {
  ApiResponse,
  CashClosing,
  CashClosingForm,
  FinanceCategory,
  FinanceSummary,
  FinanceTransaction,
  FinanceTransactionForm,
  FinanceType,
  GrossProfitPoint,
  PageResponse,
} from '@/types';

export const financeService = {
  getCategories: (activeOnly = true, type?: FinanceType) =>
    api.get<ApiResponse<FinanceCategory[]>>('/finance/categories', { params: { activeOnly, type } }),

  createCategory: (data: { name: string; type: FinanceType; active?: boolean }) =>
    api.post<ApiResponse<FinanceCategory>>('/finance/categories', data),

  getTransactions: (params: {
    page?: number;
    size?: number;
    fromDate?: string;
    toDate?: string;
    type?: FinanceType;
    categoryId?: number;
    keyword?: string;
  }) =>
    api.get<ApiResponse<PageResponse<FinanceTransaction>>>('/finance/transactions', { params }),

  createTransaction: (data: FinanceTransactionForm) =>
    api.post<ApiResponse<FinanceTransaction>>('/finance/transactions', data),

  updateTransaction: (id: number, data: FinanceTransactionForm) =>
    api.put<ApiResponse<FinanceTransaction>>(`/finance/transactions/${id}`, data),

  deleteTransaction: (id: number) =>
    api.delete<ApiResponse<void>>(`/finance/transactions/${id}`),

  getSummary: (fromDate?: string, toDate?: string) =>
    api.get<ApiResponse<FinanceSummary>>('/finance/summary', { params: { fromDate, toDate } }),

  getGrossProfitDaily: (fromDate?: string, toDate?: string) =>
    api.get<ApiResponse<GrossProfitPoint[]>>('/finance/gross-profit/daily', { params: { fromDate, toDate } }),

  getGrossProfitMonthly: (fromDate?: string, toDate?: string) =>
    api.get<ApiResponse<GrossProfitPoint[]>>('/finance/gross-profit/monthly', { params: { fromDate, toDate } }),

  getCashClosings: (params: { page?: number; size?: number; fromDate?: string; toDate?: string }) =>
    api.get<ApiResponse<PageResponse<CashClosing>>>('/finance/cash-closings', { params }),

  previewCashClosing: (businessDate?: string, openingBalance?: number) =>
    api.get<ApiResponse<CashClosing>>('/finance/cash-closings/preview', {
      params: { businessDate, openingBalance },
    }),

  closeCashDay: (data: CashClosingForm) =>
    api.post<ApiResponse<CashClosing>>('/finance/cash-closings', data),
};

File: Backend/Capstone2/src/main/java/C2SE/_1/Capstone2/service/FinanceService.java
package C2SE._1.Capstone2.service;

import C2SE._1.Capstone2.dto.CashClosingCreateDTO;
import C2SE._1.Capstone2.dto.CashClosingDTO;
import C2SE._1.Capstone2.dto.FinanceCategoryDTO;
import C2SE._1.Capstone2.dto.FinanceCategoryUpsertDTO;
import C2SE._1.Capstone2.dto.FinanceSummaryDTO;
import C2SE._1.Capstone2.dto.FinanceTransactionCreateDTO;
import C2SE._1.Capstone2.dto.FinanceTransactionDTO;
import C2SE._1.Capstone2.dto.GrossProfitPointDTO;
import C2SE._1.Capstone2.dto.PageResponse;
import C2SE._1.Capstone2.entity.FinanceType;
import C2SE._1.Capstone2.entity.User;
import org.springframework.data.domain.Pageable;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;

public interface FinanceService {

    List<FinanceCategoryDTO> getCategories(boolean activeOnly, FinanceType type);

    FinanceCategoryDTO createCategory(FinanceCategoryUpsertDTO dto);

    FinanceCategoryDTO updateCategory(Long id, FinanceCategoryUpsertDTO dto);

    void deactivateCategory(Long id);

    PageResponse<FinanceTransactionDTO> getTransactions(LocalDate fromDate,
                                                        LocalDate toDate,
                                                        FinanceType type,
                                                        Long categoryId,
                                                        String keyword,
                                                        Pageable pageable);

    FinanceTransactionDTO createTransaction(FinanceTransactionCreateDTO dto);

    FinanceTransactionDTO updateTransaction(Long id, FinanceTransactionCreateDTO dto);

    void deleteTransaction(Long id);

    FinanceSummaryDTO getSummary(LocalDate fromDate, LocalDate toDate);

    void recordAutoIncomeFromPayment(Long orderId, BigDecimal amount, LocalDateTime occurredAt, User createdBy);

    List<GrossProfitPointDTO> getGrossProfitDaily(LocalDate fromDate, LocalDate toDate);

    List<GrossProfitPointDTO> getGrossProfitMonthly(LocalDate fromDate, LocalDate toDate);

    PageResponse<CashClosingDTO> getCashClosings(LocalDate fromDate, LocalDate toDate, Pageable pageable);

    CashClosingDTO previewCashClosing(LocalDate businessDate, BigDecimal openingBalance);

    CashClosingDTO closeCashDay(CashClosingCreateDTO dto);
}
